### PR TITLE
Unify limitrange registry to be generic.

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #12207

### CODE REVIEW

**Comment by wojtek-t** [comment]
Build timed out (after 30 minutes). Marking the build as aborted.

**Comment by wojtek-t** [comment]
@k8s-bot test this please

File: pkg/registry/limitrange/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (nikhiljindal) [inline pkg/registry/limitrange/etcd/etcd.go]: 2015?

File: pkg/registry/limitrange/rest.go [hunk 2]
```diff
@@ -17,137 +17,61 @@ limitations under the License.
...
-	return rs.registry.ListPredicate(ctx, &generic.SelectionPredicate{label, field, rs.getAttrs})
+func (limitrangeStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList {
+	limitRange := obj.(*api.LimitRange)
+	return validation.ValidateLimitRange(limitRange)
 }
 
-func (rs *REST) Watch(ctx api.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error) {
-	return rs.registry.WatchPredicate(ctx, &generic.SelectionPredicate{label, field, rs.getAttrs}, resourceVersion)
+func (limitrangeStrategy) AllowUnconditionalUpdate() bool {
+	return false
```
Comment (nikhiljindal) [inline pkg/registry/limitrange/rest.go]: I think its fine to set this to true
As far as i know we support unconditional update for all resources.

**Comment by nikhiljindal** [comment]
Mostly LGTM apart from a comment about setting AllowUnconditionalUpdate() to true.
Thanks a lot for this PR. Its going to make the code a lot easier to understand.

File: pkg/registry/limitrange/etcd/etcd_test.go [hunk 3]
```diff
@@ -111,14 +110,24 @@ func TestLimitRangeCreate(t *testing.T) {
 	}
 
 	for name, item := range table {
-		fakeClient, registry := NewTestLimitRangeEtcdRegistry(t)
+		fakeClient, storage := NewTestLimitRangeStorage(t)
 		fakeClient.Data[path] = item.existing
-		err := registry.CreateWithName(ctx, key, item.toCreate)
```
Comment (nikhiljindal) [inline pkg/registry/limitrange/etcd/etcd_test.go]: We can delete CreateWithName once we have update events as well.

File: pkg/registry/limitrange/etcd/etcd.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/registry/limitrange/etcd/etcd.go]: Done.

File: pkg/registry/limitrange/etcd/etcd_test.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/registry/limitrange/etcd/etcd_test.go]: Yes - that's also part of the goal :)

File: pkg/registry/limitrange/rest.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/registry/limitrange/rest.go]: Done.

**Comment by nikhiljindal** [comment]
LGTM, thanks for the fixes!

**Comment by wojtek-t** [comment]
Jenkins failure: "Build timed out (after 30 minutes)"

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
Jenkins failure:
"Kubernetes e2e suite.Kubectl client Simple pod should support exec"
seems to be unrelated.

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
Jenkins failures because of #12467

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by a-robinson** [comment]
Needs a rebase before I can merge

**Pull request merged**